hardening/1171_postgresql_add_cache_for_created_elements

pcoello25 opened the pull request:
* Fix issue #1171 
* 100% unit tests passed:
```
[cygnus-common]: Results : Tests run: 75,    Failures: 0, Errors: 0, Skipped: 0
[cygnus-ngsi]:   Results : Tests run: 224,   Failures: 0, Errors: 0, Skipped: 0
```
* e2e (unofficial) test passed:
```
[Cygnus logs]
NGSIPostgreSQLSink[180] : [postgresql-sink] Reading configuration (cache=true)
NGSIPostgreSQLSink[194] : [postgresql-sink] PostgreSQL cache created

[First insert in PostgreSQL -> Empty Cache]
NGSIPostgreSQLSink[476] : [postgresql-sink] Persisting data at OrionPostgreSQLSink. Schema (schema1), Table (table1_car1_car), Fields ((recvTimeTs,recvTime,fiwareServicePath,entityId,entityType,attrName,attrType,attrValue,attrMd)), Values (('1475221540001','2016-09-30T07:45:40.1Z','/table1','Car1','Car','mean_speed','integer','40','[]'))
PostgreSQLCache[43] : Checking if the schema (schema1) exists
PostgreSQLCache[59] : Cache is empty
PostgreSQLBackendImpl[101] : Executing SQL query 'CREATE SCHEMA IF NOT EXISTS schema1'
PostgreSQLCache[43] : Checking if the schema (schema1) exists
PostgreSQLCache[50] : Schema (schema1) exists in Cache
PostgreSQLCache[71] : Checking if the table (table1_car1_car) belongs to (schema1)
PostgreSQLCache[77] : Table (table1_car1_car) wasn't found in the specified schema (schema1)
PostgreSQLBackendImpl[138] : Executing SQL query 'CREATE TABLE IF NOT EXISTS schema1.table1_car1_car (recvTimeTs bigint,recvTime text,fiwareServicePath text,entityId text,entityType text,attrName text,attrType text,attrValue text,attrMd text)'
PostgreSQLBackendImpl[166] : Executing SQL query 'INSERT INTO schema1.table1_car1_car (recvTimeTs,recvTime,fiwareServicePath,entityId,entityType,attrName,attrType,attrValue,attrMd) VALUES ('1475221540001','2016-09-30T07:45:40.1Z','/table1','Car1','Car','mean_speed','integer','40','[]')'
NGSISink[429] : Finishing internal transaction (f79b369a-f1ea-4daa-811b-8d767786550f)

[Second insert with same fiware-service and fiware-servicepath]
NGSIPostgreSQLSink[476] : [postgresql-sink] Persisting data at OrionPostgreSQLSink. Schema (schema1), Table (table1_car1_car), Fields ((recvTimeTs,recvTime,fiwareServicePath,entityId,entityType,attrName,attrType,attrValue,attrMd)), Values (('1475221811067','2016-09-30T07:50:11.67Z','/table1','Car1','Car','mean_speed','integer','68','[]'))
PostgreSQLCache[43] : Checking if the schema (schema1) exists
PostgreSQLCache[50] : Schema (schema1) exists in Cache
PostgreSQLCache[43] : Checking if the schema (schema1) exists
PostgreSQLCache[50] : Schema (schema1) exists in Cache
PostgreSQLCache[71] : Checking if the table (table1_car1_car) belongs to (schema1)
PostgreSQLCache[74] : Table (table1_car1_car) was found in Cache
PostgreSQLBackendImpl[166] : Executing SQL query 'INSERT INTO schema1.table1_car1_car (recvTimeTs,recvTime,fiwareServicePath,entityId,entityType,attrName,attrType,attrValue,attrMd) VALUES ('1475221811067','2016-09-30T07:50:11.67Z','/table1','Car1','Car','mean_speed','integer','68','[]')'
NGSISink[429] : Finishing internal transaction (89bff1b4-5638-4c85-86ba-1ab8f868fe9d)

[Third insert. Same schema but different tableName]
NGSIPostgreSQLSink[476] : [postgresql-sink] Persisting data at OrionPostgreSQLSink. Schema (schema1), Table (table2_car1_car), Fields ((recvTimeTs,recvTime,fiwareServicePath,entityId,entityType,attrName,attrType,attrValue,attrMd)), Values (('1475221953744','2016-09-30T07:52:33.744Z','/table2','Car1','Car','mean_speed','integer','68','[]'))
PostgreSQLCache[43] : Checking if the schema (schema1) exists
PostgreSQLCache[50] : Schema (schema1) exists in Cache
PostgreSQLCache[43] : Checking if the schema (schema1) exists
PostgreSQLCache[50] : Schema (schema1) exists in Cache
PostgreSQLCache[71] : Checking if the table (table2_car1_car) belongs to (schema1)
PostgreSQLCache[77] : Table (table2_car1_car) wasn't found in Cache
PostgreSQLBackendImpl[138] : Executing SQL query 'CREATE TABLE IF NOT EXISTS schema1.table2_car1_car (recvTimeTs bigint,recvTime text,fiwareServicePath text,entityId text,entityType text,attrName text,attrType text,attrValue text,attrMd text)'
PostgreSQLBackendImpl[166] : Executing SQL query 'INSERT INTO schema1.table2_car1_car (recvTimeTs,recvTime,fiwareServicePath,entityId,entityType,attrName,attrType,attrValue,attrMd) VALUES ('1475221953744','2016-09-30T07:52:33.744Z','/table2','Car1','Car','mean_speed','integer','68','[]')'
NGSISink[429] : Finishing internal transaction (6388d08b-c8f9-4623-80dd-13872a440ea6)

[IMPLEMENTED TEST]
Running com.telefonica.iot.cygnus.backends.postgresql.PostgreSQLCacheTest
[PostgreSQLCache.testSchemaAndTableWithSchemaCached] ---------------- Testing if a new tableName and a schemaName are in a cache with cached schemaName
[PostgreSQLCache.testSchemaAndTableWithSchemaCached] ---------  OK  - SchemaName was found in the cache but not TableName.
[PostgreSQLCache.testPersistSchemaAndTableInCache] ------------------ Persisting a new tableName and new schemaName into an empty cache
[PostgreSQLCache.testPersistSchemaAndTableInCache] -----------  OK  - Schema and Table were persisted in Cache.
[PostgreSQLCache.testSchemaAndTableInEmptyCache] -------------------- Testing if a new tableName and new schemaName are in an empty cache
[PostgreSQLCache.testSchemaAndTableInEmptyCache] -------------  OK  - TableName and SchemaName weren't found in the empty Cache.
[PostgreSQLCache.testIfPersistDuplicateTableNames] ------------------ Persisting a tableName and schema into a cache with both values already cached.
[PostgreSQLCache.testIfPersistDuplicateTableNames] -----------  OK  - Values weren't duplicated.
[PostgreSQLCache.testSchemaAndTableWithBothValuesCached] ------------ Testing if a tableName and a schemaName are in a cache with both values cached
[PostgreSQLCache.testSchemaAndTableWithBothValuesCached] -----  OK  - TableName and SchemaName were found in the Cache.
[PostgreSQLCache.testPersistOnlyTableInCache] ----------------------- Persisting a new tableName into a schema already created in cache
[PostgreSQLCache.testPersistOnlyTableInCache] ----------------  OK  - Table was persisted in the cached schema.
Tests run: 6, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.006 sec
```

frbattid commented on cygnus-common/src/main/java/com/telefonica/iot/cygnus/backends/postgresql/PostgreSQLCache.java:
```diff
@@ -0,0 +1,101 @@
+    } // persistSchemaInCache
+    
+    public void persistTableInCache(String schemaName, String tableName) {
+        ArrayList<String> tableNames = cache.get(schemaName);
```
You are assuming the schema is already in the cache... It may not.

pcoello25 replied:
Yes, my fault, forgot to check.
Fixed in ef962c8

frbattid replied:
Instead of doing this, which loops twice (one in order to check if the schema exists, another in order to put the table), try the following: iterate on the schemas, and once found add the table. If not found, nothing is done.

frbattid commented on cygnus-common/src/test/java/com/telefonica/iot/cygnus/backends/postgresql/PostgreSQLCacheTest.java:
```diff
@@ -0,0 +1,176 @@
+/**
+ * Copyright 2014 Telefonica Investigación y Desarrollo, S.A.U
```
2016 ;)

pcoello25 replied:
Fixed in ef962c8

frbattid commented on cygnus-common/src/test/java/com/telefonica/iot/cygnus/backends/postgresql/PostgreSQLCacheTest.java:
```diff
@@ -0,0 +1,176 @@
+     * @throws Exception
+     */
+    @Before
+    public void setUp() throws Exception {
```
Instead of using `setUp`, create the required object within each test method.

pcoello25 replied:
Fixed in ef962c8

frbattid commented on cygnus-ngsi/src/main/java/com/telefonica/iot/cygnus/sinks/NGSIPostgreSQLSink.java:
```diff
@@ -48,6 +49,9 @@
     private PostgreSQLBackendImpl persistenceBackend;
+    // Parameter 'cache' and cache <schema,[table,table,table..]>
+    private boolean enableCache;
+    private PostgreSQLCache postgreCache;
```
This should not be necessary here, right?

frbattid commented on cygnus-ngsi/src/main/java/com/telefonica/iot/cygnus/sinks/NGSIPostgreSQLSink.java:
```diff
@@ -63,6 +67,14 @@ public NGSIPostgreSQLSink() {
+     * Gets the PostgreSQL cache. It is protected due to it is only required for testing purposes.
+     * @return The PostgreSQL cache state
+     */
+    protected boolean getCache() {
```
getEnableCache instad of getCache

pcoello25 replied:
Fixed in ef962c8

frbattid commented on cygnus-ngsi/src/test/java/com/telefonica/iot/cygnus/sinks/NGSIPostgreSQLSinkTest.java:
```diff
@@ -858,6 +877,41 @@ public void testBuildTableNameLengthDataModelByAttribute() throws Exception {
     } // testBuildTableNameLengthDataModelByAttribute
     
+        /**
+     * [NGSIPostgreSQLSink.configure] -------- cache can only be 'true' or 'false'.
```
This seems to be missaligned.

pcoello25 replied:
Fixed in ef962c8

frbattid commented on doc/cygnus-ngsi/flume_extensions_catalogue/ngsi_postgresql_sink.md:
```diff
@@ -239,7 +239,7 @@ Coming soon.
 | batch_timeout | no | 30 | Number of seconds the batch will be building before it is persisted as it is. |
 | batch_ttl | no | 10 | Number of retries when a batch cannot be persisted. Use `0` for no retries, `-1` for infinite retries. Please, consider an infinite TTL (even a very large one) may consume all the sink's channel capacity very quickly. |
-
+| backend.enable_cache | yes | N/A | <i>true</i> or <i>false</i>, <i>true</i> enable the creation of a Cache, <i>false</i> disable the creation of a Cache |
```
enable -> enables
disable -> disables
Use a final dot `.` in the phrase

pcoello25 replied:
Fixed in ef962c8

frbattid commented on doc/cygnus-ngsi/installation_and_administration_guide/ngsi_agent_conf.md:
```diff
@@ -169,6 +169,8 @@ cygnusagent.sinks.postgresql-sink.batch_size = 100
 cygnusagent.sinks.postgresql-sink.batch_timeout = 30
 # number of retries upon persistence error
 cygnusagent.sinks.postgresql-sink.batch_ttl = 10
+# true enables cache, false disables cache
```
The agent configuration template must be updated as well.

pcoello25 replied:
Fixed in ef962c8

pcoello25 commented:
Fixed in 85f172b. I've modified the loop with the schema iteration (Call the method checking if the schema exists and iterate over the list of schemas is the same. We avoid the repeated logs as you suggest)

frbattid commented on cygnus-common/src/main/java/com/telefonica/iot/cygnus/backends/postgresql/PostgreSQLBackendImpl.java:
```diff
@@ -50,9 +51,14 @@
+        if (enableCache) {
+            cache = new PostgreSQLCache();
+            LOGGER.info("PostgreSQL cache created succesfully");
+        } // if
```
White line required after `{`.

pcoello25 replied:
Fixed in d37ef84

frbattid commented on cygnus-common/src/main/java/com/telefonica/iot/cygnus/backends/postgresql/PostgreSQLBackendImpl.java:
```diff
@@ -200,8 +218,8 @@ private void closePostgreSQLObjects(Connection con, Statement stmt) throws Excep
          */
-        public PostgreSQLDriver(String postgresqlHost, String postgresqlPort,
-                String postgresqlDatabase, String postgresqlUsername, String postgresqlPassword) {
+        public PostgreSQLDriver(String postgresqlHost, String postgresqlPort,String postgresqlDatabase, 
```
White space required after comma at `postgresqlPort,String postgresqlDatabase`

pcoello25 replied:
Fixed in d37ef84

frbattid commented on cygnus-ngsi/conf/agent_ngsi.conf.template:
```diff
@@ -214,6 +214,8 @@ cygnusagent.sinks.postgresql-sink.type = com.telefonica.iot.cygnus.sinks.NGSIPos
 # number of retries upon persistence error
 # cygnusagent.sinks.postgresql-sink.batch_ttl = 10
+# true enables cache, false disables cache
+cygnusagent.sinks.postgresql-sink.backend.enable_cache = false
```
This is an optional parameter, since it has a defulat value (`false`).

pcoello25 replied:
Fixed in d37ef84

frbattid commented on cygnus-ngsi/src/main/java/com/telefonica/iot/cygnus/sinks/NGSIPostgreSQLSink.java:
```diff
@@ -48,6 +48,8 @@
     private String postgresqlPassword;
     private boolean rowAttrPersistence;
     private PostgreSQLBackendImpl persistenceBackend;
+    // Parameter 'cache' and cache <schema,[table,table,table..]>
```
Leftover comment?

pcoello25 replied:
Fixed in d37ef84

frbattid commented on doc/cygnus-ngsi/flume_extensions_catalogue/ngsi_postgresql_sink.md:
```diff
@@ -239,7 +239,7 @@ Coming soon.
 | batch_timeout | no | 30 | Number of seconds the batch will be building before it is persisted as it is. |
 | batch_ttl | no | 10 | Number of retries when a batch cannot be persisted. Use `0` for no retries, `-1` for infinite retries. Please, consider an infinite TTL (even a very large one) may consume all the sink's channel capacity very quickly. |
-
+| backend.enable_cache | yes | N/A | <i>true</i> or <i>false</i>, <i>true</i> enables the creation of a Cache, <i>false</i> disables the creation of a Cache. |
```
This is wrong, isn't it? It is not a mandatory parameter since it has a default value (`false`).

pcoello25 replied:
Fixed in d37ef84

frbattid commented on doc/cygnus-ngsi/installation_and_administration_guide/ngsi_agent_conf.md:
```diff
@@ -169,6 +169,8 @@ cygnusagent.sinks.postgresql-sink.batch_size = 100
 # number of retries upon persistence error
 cygnusagent.sinks.postgresql-sink.batch_ttl = 10
+# true enables cache, false disables cache
+cygnusagent.sinks.postgresql-sink.backend.enable_cache = false
```
This is an optional parameter, since it has a defulat value (`false`).

pcoello25 replied:
Fixed in d37ef84
